### ISSUE TITLE
Unable to deploy shared load balancer on different paths in same domain

### ISSUE THREAD

**Issue by ds29** [aws/copilot-cli#2527]
Hi, first and foremost, thank you to the Copilot Team. This CLI is amazing and has saved me weeks worth of learning the in/outs of deploying a modern container based stack.

That said, I'm encountering a few issues when trying to deploy a load balancer on two different paths. My application is a simple Rails API backed by Postgres and uses a Next JS frontend. I have one environment in copilot: `staging`. I have two services defined, `backend` (the Rails API with Aurora add-on) and `frontend` (the Next JS app).

I'm using version 1.8.1 of copilot so that I could make use of the custom domain alias. My use case is to have my application served from `app.example.com`, so my manifest defines the `http.alias` as `app.example.com` for both services.

Following recommendations I read on the gitter.im room, I deployed `backend` service listening on the `/api/v1` path. Upon deployment, my service says it's available at `https://app.example.com//api/v1`. Note the extra double `//`, it seems that copilot is adding this extra `/` on it's own. When I attempted to correct that by using only `api/v1' for the path in my manifest, copilot alerted that this was invalid configuration and required the leading `/`. 🤷‍♂️

That issue aside, I then proceeded to deploy my `frontend` service listening on the root path `/` using the `http.alias` of `app.example.com`. That produces an error when it attempts to add `A-record` to the hosted zone.

Removing the `http.alias` allows the `frontend` service to be deployed to the default `frontend.staging.{app}.{domain}` path but doesn't allow me to run my application as desired.

For reference, I'm sharing my two service manifests (sanitized) below. Any help/guidance you could provide is GREATLY appreciated.

```yaml
# backend/manifest.yml
name: backend
type: Load Balanced Web Service

http:
  path: '/api/v1'
  healthcheck: '/api/v1/health_check'

build
  build: api/Dockerfile
  port: 3000

cpu: 256
memory: 512
count: 1
exec: true

secrets:
  RAILS_MASTER_KEY: RAILS_MASTER_KEY

environments:
  staging:
    count: 1
    http:
      alias: app.example.com
    secrets:
      RAILS_MASTER_KEY: /copilot/myapp/staging/secrets/RAILS_MASTER_KEY
    variables:
      RAILS_ENV: 'staging'
```

```yaml
# frontend/manifest.yml
name: frontend
type: Load Balanced Web Service

http:
  path: '/'
  healthcheck: '/api/health'

image:
  # Docker build arguments. For additional overrides: https://aws.github.io/copilot-cli/docs/manifest/lb-web-service/#image-build
  build: app/Dockerfile
  port: 3001

cpu: 256
memory: 512
count: 1
exec: true

environments:
  staging:
    count: 1
    http:
      alias: app.example.com
```

**Comment by iamhopaul123** [comment]
Hello @ds29. Sorry for the churn. I tried to reproduce the steps but it worked well for me. Here's what I did:

1. I created an app using domain `archercopilot.com` with the app name to be `demo` and an env named `test` and an LB service named `frontend`.
2. Part of the LB service manifest is shown as below:
``` yaml
http:
  # Requests to this path will be forwarded to your service.
  # To match all requests you can use the "/" path.
  path: 'frontend'
  # You can specify a custom health check path. The default is "/".
  # healthcheck: '/'
  alias: archercopilot.com
```
Then after the deployment, it showed ```✔ Deployed frontend, you can access it at https://archercopilot.com/frontend.```

Technically you should have the path to be `api/v1` without the leading "/", and I'm not sure why it shows `api/v1' for the path in my manifest, copilot alerted that this was invalid configuration and required the leading /`. Can you paste the whole error message we throw here?

Also for the other issue, is `example.com` the same domain when you created the app? If not then Copilot is not allowed to automatically help you to validate the certificate as described [here](https://github.com/aws/copilot-cli/pull/2482/files#diff-cfa49eb43e38e53f634441c85263ef902825531e7ba7e2f773cfa997b4c3ef9aR15-R25) (sorry we'll release this doc change  showing this limitation as well as manifest alias validation in our next release!) However, hopefully after https://github.com/aws/copilot-cli/pull/2413 we'll allow users to import any validated certificate so as to workaround this limitation. Thank you!

**Comment by ds29** [comment]
@iamhopaul123 Thank you for the quick response! After reading through your message, I realized my confusion on the leading `/` issue. The error I was getting was for the `healthcheck`, which requires leading `/` on the path. However, the `path` for load balancer `http` does NOT expect the leading `/`. With that corrected, I'm able to deploy my `backend` service to the correct url.

However, on the other issue, I'm still not able to deploy both services using the `app.example.com` domain alias. I actually have two applications, one running on `example.com` and this one running on `app.example.com`. I'm using copilot for both but they are in separate VPCs, etc.  For the `app.example.com`, I initially tried to init `myapp` using `copilot app init --domain app.example.com`. That complained about not having a hosted zone for `app.example.com`. So, I created a new hosted zone for that. However, it still did not work, so I deleted the app and created a new stack using `copilot app init --domain example.com`.  This worked to get me application running at `app.example.com` but now still produces this error on the deployment of my second service (`frontend`).

Here is the error I get (actual app name replaced with `myapp`):
```
Updating the infrastructure for stack myapp-staging-frontend                 [update rollback complete]  [174.2s]
  The following resource(s) failed to update: [EnvControllerAction].
  - Update your environment's shared resources                                       [update rollback complete]  [109.0s]
    The following resource(s) failed to update: [HTTPSListener, CustomDoma
    inAction].
    - Request and validate an ACM certificate for your domain                        [delete complete]           [1.8s]
    - Add an A-record to the hosted zone for the domain alias                        [update complete]           [90.2s]
  - An ECS service to run and maintain your tasks in the environment cluster         [not started]
  - An ECS task definition to group your containers and run them on ECS              [not started]
✘ deploy service: stack myapp-staging-frontend did not complete successfully and exited with status UPDATE_ROLLBACK_COMPLETE
```

Following the logs related to the `EnvControllerAction`, the lambda produces this error message:
```
2021-06-27T01:21:42.522Z	9b93de15-aaf4-47fb-aea8-2b4db937ae5d	INFO	Caught error ResourceNotReady: Resource is not in the state stackUpdateComplete.
```

I'm not sure if this is a bug or if my order of operations caused some unintended behavior in my setup.

**Comment by iamhopaul123** [comment]
> For the app.example.com, I initially tried to init myapp using copilot app init --domain app.example.com. That complained about not having a hosted zone for app.example.com.

If you own `example.com` then I think you would need to create the subdomain on your own in order to make it work. However, I don't think this is ideal since now you can just use `alias`.

> This worked to get me application running at app.example.com but now still produces this error on the deployment of my second service (frontend).

So sorry about it. Right now for `v1.8.1` we are having [this issue](https://github.com/aws/copilot-cli/issues/2504
) but should be fixed by #2513, which will be included in our next release very soon! Please stay tuned. Thank you!

**Comment by ds29** [comment]
> So sorry about it. Right now for `v1.8.1` we are having [this issue](https://github.com/aws/copilot-cli/issues/2504) but should be fixed by #2513, which will be included in our next release very soon! Please stay tuned. Thank you!

No worries, thanks so much for this awesome project and the quick responses! Looking forward to the next update!

**Comment by iamhopaul123** [comment]
Heya we fixed this in our latest release [v1.8.2](https://github.com/aws/copilot-cli/releases/tag/v1.8.2). Thank you for reporting this issue ❤️

**Comment by ds29** [comment]
@iamhopaul123 I've updated to `1.8.2` but still can't get my two services to deploy to the same subdomain. I deploy my backend service first and then the frontend service.  Backend deploys to `app.example.com` and then frontend fails with:

```
⠙ Proposing infrastructure changes for stack myapp-staging-frontend ✔ Proposing infrastructure changes for stack myapp-staging-frontend
- Updating the infrastructure for stack myapp-staging-frontend                 [update rollback complete]  [213.0s]
  The following resource(s) failed to update: [EnvControllerAction].
  - Update your environment's shared resources                                       [update rollback complete]  [162.1s]
    The following resource(s) failed to update: [HTTPSListener, CustomDoma
    inAction].
    - Request and validate an ACM certificate for your domain                        [delete complete]           [3.2s]
    - Add an A-record to the hosted zone for the domain alias                        [update complete]           [133.2s]
      Resource update cancelled
  - An ECS service to run and maintain your tasks in the environment cluster         [not started]
  - An ECS task definition to group your containers and run them on ECS              [not started]
✘ deploy service: stack myapp-staging-frontend did not complete successfully and exited with status UPDATE_ROLLBACK_COMPLETE
```

In an effort to resolve that, I attempted to remove my frontend service and try again with a fresh setup for that service which resulted in the same failure. I'm not sure if there's some leftover remnants from the bug or where to go from here.

Any ideas?

**Comment by ds29** [comment]
If it helps, these are the logs output from the `EnvControllerFunction` lambda:

```
START RequestId: 53f04b9b-1bcc-45dd-a82f-d8c8fe84fde2 Version: $LATEST
--
2021-06-30T20:57:33.527Z	53f04b9b-1bcc-45dd-a82f-d8c8fe84fde2	INFO	Caught error ResourceNotReady: Resource is not in the state stackUpdateComplete.
2021-06-30T20:57:33.527Z	53f04b9b-1bcc-45dd-a82f-d8c8fe84fde2	INFO	Responding FAILED for physical resource id: envcontoller/myapp-staging/frontend
END RequestId: 53f04b9b-1bcc-45dd-a82f-d8c8fe84fde2
REPORT RequestId: 53f04b9b-1bcc-45dd-a82f-d8c8fe84fde2	Duration: 121275.07 ms	Billed Duration: 121276 ms	Memory Size: 512 MB	Max Memory Used: 94 MB	Init Duration: 374.39 ms
START RequestId: 58713c40-f328-40a0-bebb-c90f9af5bc54 Version: $LATEST
END RequestId: 58713c40-f328-40a0-bebb-c90f9af5bc54
REPORT RequestId: 58713c40-f328-40a0-bebb-c90f9af5bc54	Duration: 105.93 ms	Billed Duration: 106 ms	Memory Size: 512 MB	Max Memory Used: 94 MB
```

**Comment by iamhopaul123** [comment]
I think most likely it is the leftover remnants. Would you clean up the ACM certs previously created by Copilot? Also I was wondering why did you specify the domain name for you backend service since backend service is not supposed to be accessed through public internet?

**Comment by ds29** [comment]
One more update for better clarity on my particular setup. I have two copilot applications in separate git repos.

I initialized the first application using `copilot init --domain example.com` with a name of `marketing`. It has the default VPC setup in one environment `staging`.

The second application (the one that's failing) was initialized using `copilot init --domain example.com` with a name of `myapp`. It also has the default VPC setup in one environment `staging`.

'marketing` has one load balanced web service: `frontend` and uses `http.alias` of `example.com`.
`myapp` has two load balance web services: `backend` and `frontend` and I'm attempting to deploy both services using the `http.alias` of `app.example.com`.

**Comment by iamhopaul123** [comment]
Hmm but it is not possible to deploy both load balanced web services using the same alias. Why do you want to use the same alias 🤔 If you do that what service do you expect to handle the traffic coming to `app.example.com`?

So the `http.alias` is to specify the alias you use for this deployed service, instead of modifying the root domain you specified when you create the app.

**Comment by ds29** [comment]
> Also I was wondering why did you specify the domain name for you backend service since backend service is not supposed to be accessed through public internet?

Yes, it's a good question. It's poor terminology on my part but `backend` service is actually a load balance publicly accessible API.

**Comment by ds29** [comment]
> Hmm but it is not possible to deploy both load balanced web services using the same alias. Why do you want to use the same alias 🤔 If you do that what service do you expect to handle the traffic coming to `app.example.com`?

In my setup, copilot has correctly configured the load balancer in the environment using the `path`. I have my `backend` service path set to `services` so any request to `example.com/services` is routed to `backend` services. Since this service was created first, the LB rule is first.  Then, my `frontend` service uses `/` for the LB and would route any request that isn't `services` to the `frontend` service.

I could do this routing using nginx but I thought that copilot would handle this at the LB.

**Comment by iamhopaul123** [comment]
Gotcha that totally makes sense to me. I'll try to reproduce the error and see how we can address this use case. Thank you for reporting the issue.

**Comment by ds29** [comment]
If a picture helps, this is the desired configuration.
![LB Layout](https://user-images.githubusercontent.com/7771/124038532-41c7c180-d9cf-11eb-86d2-dd000c86712c.png)

That said, if there's a better way / more common way to achieve this using the existing copilot features. Let me know. My only reason for this approach was to avoid having to maintain / setup nginx proxy containers. In the end, I just want traffic routing to the various URLS to go the proper destinations.  If this is better achieved in a single copilot application, I can go that route too.

Thanks again for the awesome work from this team and many more thanks for the helpful guidance!

**Comment by iamhopaul123** [comment]
Hello @ds29. Thank you so much for the reply. The diagram you included is awesome! I think right now if I am understanding correctly Copilot `v1.8.2` is possible to address your need. You might just need to clean up the leftover remnants caused by the previous version. So here's what I did:

1. `copilot app init --domain penghaoh.com` to create an app.
2. `copilot env init --name test` to create an test env.
3. `copilot svc init -n frontend` to create frontend service with path to be `frontend` and alias to be `penghaoh.com`.
4. `copilot svc init -n backend` to create backend service with path to be `backend` and alias to be `penghaoh.com`.
5. `copilot svc deploy` to deploy the backend service first and then frontend service.

After the above steps, both LB services can be accessed through `https://penghaoh.com/frontend` and `https://penghaoh.com/backend`. Let me know if you have any questions!

**Comment by ds29** [comment]
@iamhopaul123 Thanks! I'm going to give it another try. Regarding this bit:

> You might just need to clean up the leftover remnants caused by the previous version.

What would need to be cleaned up? Is there a way I can do that without tearing down the whole env and starting fresh?

**Comment by iamhopaul123** [comment]
@ds29 Usually the ACM certs and route53 records are the only things. Oooh i think i figured it out. It's because we didn't update the env version for the latest release, and that's the reason why the fix was not populated to your existing env. Really sorry about that. For the workaround would you mind trying the following step?

1. Copy your env template and save it as a file locally
2. Modify the version of the template from `v1.4.0` to `v1.3.0`, then update the env stack with the updated template using the CFN console, with the same setting as the previous deployment.
3. Then in your next service deployment Copilot shall help you to upgrade the environment to the latest version with the new fix. In this way you don't need to tear down the whole env.

**Comment by ds29** [comment]
@iamhopaul123 Thanks again. Dropping the stack down to `v1.3.0` seemed to make a difference but I suspect that I may still have some extra leftovers on Route 53 records. I wasn't able to get my current stack working.

However, I decided to simplify and put together a sample app to see if copilot would do what I was trying to do. This repo is a very simple set of node servers that mimic my desired setup: https://github.com/ds29/aws-copilot-demo

I can confirm that copilot works as desired using this repo on a fresh setup. So, my current stack isn't working as I'd like but now I no it's possible.  Hopefully this repo might help others who happen across this issue.

Thanks again for your help!

**Comment by iamhopaul123** [comment]
Hello @ds29 sorry for the late reply. Glad to see Copilot finally is able to build something as desired! However, I would be still curious to see why the previous stack won't work even after the env stack template upgrade. Let me know if you need any help on your project!

**Comment by ds29** [comment]
@iamhopaul123 Thanks, I ended up removing the stack which seemed to clear all the old attempts. The fresh start approach worked for me. I suspect there were multiples of the lambda functions lying around which likely caused confusion to copilot when it saw existing resources. If I had attempted to remove all of those first, I expect that I wouldn't have needed to take the teardown approach.

That said, in my case, I actually changed my configuration slightly so the teardown was useful.
